compiler/damlc/daml-package-config/src/DA/Daml/Package/Config.hs

```diff
@@ -121,6 +146,52 @@ withPackageConfig projectPath f = do
     pkgConfig' <- overrideSdkVersion pkgConfig
     f pkgConfig'
 
+-- Traverses up the directory tree from current project path and returns the project path of the "nearest" project.yaml
+-- Stops at root, but also won't pick any files it doesn't have permission to search
+findMultiPackageConfig :: ProjectPath -> IO (Maybe ProjectPath)
+findMultiPackageConfig projectPath = do
+  filePath <- canonicalizePath $ unwrapProjectPath projectPath
+  flip loopM filePath $ \path -> do
+    hasMultiPackage <- doesFileExist $ path </> multiPackageConfigName
+    if hasMultiPackage
+      then pure $ Right $ Just $ ProjectPath path
+      else
+        let newPath = takeDirectory path
+        in pure $ if path == newPath then Right Nothing else Left newPath
+
+canonicalizeMultiPackageConfigIntermediate :: ProjectPath -> MultiPackageConfigFieldsIntermediate -> IO MultiPackageConfigFieldsIntermediate
+canonicalizeMultiPackageConfigIntermediate projectPath (MultiPackageConfigFieldsIntermediate (MultiPackageConfigFields packagePaths) multiPackagePaths) =
+  withCurrentDirectory (unwrapProjectPath projectPath) $ do
+    MultiPackageConfigFieldsIntermediate
+      <$> (MultiPackageConfigFields <$> traverse canonicalizePath packagePaths)
+      <*> traverse canonicalizePath multiPackagePaths
+
+-- Fix for single argument that checks for loops over that single argument in the recursion
```

Comment:
I changed the comment, but I wasn't fully convinced by yours either tbh, starting off with `checks` implies this is a predicate function, rather than something that produces a result

Reply:
Yeah it's a bit of a pain to describe - how about

```
Runs an IO action that takes its own fixpoint, but aborts if the IO action would recurse infinitely by passing itself the same argument over and over.
```